### PR TITLE
Low-level API Poisson equation

### PR DESCRIPTION
**Pull request by amartinhuertas**
Hi @fverdugo, @santiagobadia,

after some final touches (see issue https://github.com/gridap/Tutorials/issues/60 for more details), I think the Low-Level tutorial is ready to merge! 

The only thing which is missing is the following observation: 

https://github.com/gridap/Tutorials/blob/dev-tutorials/src/poisson_dev_fe.jl#L368 

for which I do not have an answer yet. We can leave it there, remove it, or try to find the reason, answer and solve it if necessary.

What do you think?

### CODE REVIEW

**Comment by fverdugo** [comment]
Hi @amartinhuertas excellent job! Perhaps we can discuss it face-to-face since there is a lot of (good) marterial.

Some comments to be discussed
- One can use a number of integration points different from the number of shape functions. It will help to see which is the axis for points and which for local dofs.
- The `Debugger` can also be called from the REPL. It would be nice to check if it can be also used from a notebook.
- export `FEBasis`, `BasisStyle`, `TestBasis` and `TrialBasis` from `FESpaces` (but not from `Gridap`) 
-  As you have noticed, we don't yet have 100% allocation free loops in some cases. I think it is perhaps related with this issue https://github.com/JuliaLang/julia/issues/35800 This is something we want to solve at some point.
- The actual purpose of `Broadcasting` is allowing to write `lazy_map(Broadcasting(*),a,b)` instead of `lazy_map((ai,bi)->ai.*bi,a,b)`. The main reason is syntactic sugar (even though there are a number of dispatches specialized for Broadcasting for performance reasons.)
- Keep in mind that, when executed in a notebook, only the return value of the last line is displayed
- Use the Gridap builtin `print_op_tree` instead of `print_lazy_array_type_parameters`. This also will simplify the reading since you don't need to expose so many details of `LazyArray` (whics is quite overwhelming). 
- `lazy_map(k,a,b)` is semantically (conceptually) equivalent to `map(k,a,b)` for a callable object `k
- Objects `k` such that  `isa(k,Map)==true` are callable. i.e. `k(x...)` is equivalent to `evaluate(k,x...)`
- `∑(integrate((u-ufₕ)*(u-ufₕ),Qₕ))` ->  `∑( ∫( (u-ufₕ)*(u-ufₕ) )Qₕ )` even more unicode-ish

**Comment by amartinhuertas** [comment]
Hi @fverdugo ... I went over all your comments in regards to this PR. See https://github.com/gridap/Tutorials/pull/71#issuecomment-777632964 for more details. 

I had the following issues:

* I tried to put consecutive lines which I want in different code cells by putting a blank line among them, but it did not work. Do u know how to do this?
* The usage of `print_op_tree` causes an error in the CI system.  https://github.com/gridap/Tutorials/runs/2084255358?check_suite_focus=true#step:6:1 Have you ever seen this error?
* I could not test whether `Debugger` works in JNs. My local installation is broken, and I do not want to fix it now because I am lecturing with JNs. Can you try this and comment at the begnining of the tutorial if this possibility exists?
* The other points which are under the pending header are points which I did not actually know what you pretended I do with them.

**Comment by fverdugo** [comment]
Thanks! I'll take a look.

**Comment by fverdugo** [comment]
> I tried to put consecutive lines which I want in different code cells by putting a blank line among them, but it did not work. Do u know how to do this?

Simply by adding an empty comment line between them. I.e.,
```julia
DomainStyle(Qₕ) == ReferenceDomain()
#
DomainStyle(Qₕ) == PhysicalDomain()
```

> I could not test whether Debugger works in JNs. My local installation is broken, and I do not want to fix it now because I am lecturing with JNs. Can you try this and comment at the begnining of the tutorial if this possibility exists?

To simplify things (for us and for the reader), I would write the tutorial assuming that the reader will do it by executing the cells in the notebook (just like the other tutorials). If you need to show intermediate results, you can split the cells as commented above.  I would recommend to use a debugger just as an optional step. 

> The usage of print_op_tree causes an error in the CI system. https://github.com/gridap/Tutorials/runs/2084255358?check_suite_focus=true#step:6:1 Have you ever seen this error?

I have never seen this error, but by looking at the implementation of `print_op_tree` in Gridap it seems that there is a bug:

https://github.com/gridap/Gridap.jl/blob/3e3b006b5046b628449c0053af4e5c82c5063111/src/Arrays/PrintOpTrees.jl#L34

It should be `stdout` instead of `stdin`, right?

> The other points which are under the pending header are points which I did not actually know what you pretended I do with them.

The issue about spurious memory allocation still has to be further investigated...

**Comment by fverdugo** [comment]
Hi @amartinhuertas! please find my replay to your last questions above ☝️

**Comment by fverdugo** [comment]
Regarding the bug in `print_op_tree` https://github.com/gridap/Gridap.jl/pull/563

**Comment by amartinhuertas** [comment]
> Simply by adding an empty comment line between them. I.e.,

Ok, I do not know why, but I though this was a dirty hack, and I was expecting a more sophisticated solution (e.g., an special markdown tag). I already followed this approach in 83f986ded6791af690c2acdf3e9942f6fb943381. 

> It should be stdout instead of stdin, right?

Yes. it should be stdout. I am now leveraging your bug-fix in this branch. Let us see whether this solves the problem. 

> The issue about spurious memory allocation still has to be further investigated...

Yes, but apart from this, there are still two additional comments without solution yet.

**Comment by amartinhuertas** [comment]
> Regarding the bug in print_op_tree gridap/Gridap.jl#563

This solves the issue in the tutorials CI ! Great!

**Comment by fverdugo** [comment]
> export FEBasis, BasisStyle, TestBasis and TrialBasis from FESpaces (but not from Gridap)

The definition of `FEBasis` will change in Gridap v0.16. `FEBasis` will be an abstract type, with `SingleFieldFEBasis` and `MultiFieldFEBasis` specializations. I expect to register Gridap v0.16 during this week. They will be exported from the corresponding modules.

> The actual purpose of Broadcasting is allowing to write lazy_map(Broadcasting(*),a,b) instead of lazy_map((ai,bi)->ai.*bi,a,b). The main reason is syntactic sugar (even though there are a number of dispatches specialized for Broadcasting for performance reasons.)

This is just a minor comment.

**Comment by amartinhuertas** [comment]
Ok ... so the pull request can be accepted, right?

**Comment by fverdugo** [comment]
> Ok ... so the pull request can be accepted, right?

By looking at the Manifest, it seems that this PR relies on an unregistered version of Gridap. I'd wait until we release Gridap 0.15.3 with the patch related to the print_op_tree and change the Manifest here accordingly.

**Comment by amartinhuertas** [comment]
>By looking at the Manifest, it seems that this PR relies on an unregistered version of Gridap.

Yeah, true. I should remove the Manifest.toml from the github repo. 

> I'd wait until we release Gridap 0.15.3 with the patch related to the print_op_tree and change the Manifest here accordingly.

Ok. Let us wait until them. Are we waiting for a particular event in order to release Gridap 0.15.3? If not, I can help to prepare this new release.

**Comment by fverdugo** [comment]
> Ok. Let us wait until then. Are we waiting for a particular event in order to release Gridap 0.15.3? If not, I can help to prepare this new release.

Perfect if you can release this!

The only part missing is to add a couple of entries in the NEWS.md related with the last PRs, move to 0.15.3 in the Project.toml, check that all tests pass and release. See also https://github.com/gridap/Gridap.jl/wiki/How-to-register-a-julia-package#specific-steps-to-registering-gridap

**Comment by amartinhuertas** [comment]
> The only part missing is to add a couple of entries in the NEWS.md related with the last PRs

Looking at the PRs accepted since the last release (0.15.2), I see (ordered by PR id, not chronologically): #553, #555, #556, #559, #563, #564. If this is the case, we should add 6 entries to the NEWS.md file (not actually a couple), right? On the other hand ... are all these minor release patches?

**Comment by fverdugo** [comment]
Yes, sorry. We usually only include the most relevant changes in NEWS.md, it **is not needed to include all PRs**. 

In this case, I would mention that 1) we have fixed an bug in print_op_tree and  2) that we the cell map in meshes of linear simplices is now an array of affine maps, which allows one to compute laplacians of the shape functions among other things.

**Comment by amartinhuertas** [comment]
Ok, thanks! PR requested in https://github.com/gridap/Gridap.jl/pull/566

**Comment by amartinhuertas** [comment]
@fverdugo ... all tests passed with Gridap 0.15.3 (new release). I think the PR is ready to merge.

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (fverdugo) [inline .vscode/launch.json:1]: Perhaps we can add this file to `.gitignore`

File: Project.toml [hunk 2]
```diff
@@ -15,7 +16,6 @@ Printf = "de0858da-6303-5e67-8744-51eddeeeb8d7"
 Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 
 [compat]
-Gridap = "0.15"
```
Comment (fverdugo) [inline Project.toml:18]: Please, do not remove Gridap from the compat section. Use Gridap 0.15.3 since this is the version that has the bugfix we need.

**Review by fverdugo — CHANGES_REQUESTED** [review]
A couple of minor comments and we are ready to go!

File: .vscode/launch.json (hunk 1 above)
Comment (amartinhuertas) [reply .vscode/launch.json]: It was already in .gitignore. Possible added accidentally into the repository before was in gitignore. I have removed the file.

File: Project.toml (hunk 2 above)
Comment (amartinhuertas) [reply Project.toml:18]: Ok, Added. But I was not aware that I had removed it ... perhaps it was the package manager automatically or something like that?

**Comment by fverdugo** [comment]
Hi @amartinhuertas. Thanks for the changes.

Just a question, what is `dev/Gridap` for? Do we need it?

**Comment by amartinhuertas** [comment]
> Just a question, what is dev/Gridap for? Do we need it?

I dont think so. Removed. You have 100 eyes!

**Review by fverdugo — APPROVED** [review]

**Pull request merged**